src/compiler/types.ts

```diff
@@ -3613,6 +3613,9 @@ namespace ts {
         BooleanLike = Boolean | BooleanLiteral,
         EnumLike = Enum | EnumLiteral,
         ESSymbolLike = ESSymbol | UniqueESSymbol,
+        VoidLike = Void | Undefined,
+        /* @internal */
+        DisjointDomains = NonPrimitive | StringLike | NumberLike | BooleanLike | ESSymbolLike | VoidLike | Null,

```

Comment:
Object types can also have disjoint domains, eg 
```ts
type A = { x: string };
type B = { x: number };
type C = { x: boolean };
function f5<T extends A | B>(x: T & (B | C)) {
    const y: B = x; // expected to work, currently does not, doesn't work with this PR
}
```
(important if you just consider each of those types of `x` as different tags for tagged unions, eg `"a"`, `"b"`, and `"c"`)

What we have here is certainly better, but definitely still not quite complete.

Reply:
> Also, we still allow stuff like string & { __tag__?: void } even though that's technically an impossible type.

I feel like you can justify that by claiming that `{}` types indicate structure, but no top level domain (which is mostly true - `{ toString(): string }` matches primitives AFAIK), so it's just string-augmented-with-extras. So it doesn't feel so bad.

Also, it's probably feasible for object types - we do compare them based on that structure, after all, and we're even within that comparison when this simplifier/inliner gets called. And it's not like `{ x: number } & { x: string }` should simplify to `never` - it should be `{ x: never }` (and `{ x: number } | { x: never }` should be `{ x: number }`) - it's exactly what we'd do when we actually compare the properties (or asked for `obj.x`), but not deferred until access so that the top-level union/intersection relationship works out (and this is all about doing union/intersection simplification more eagerly to make comparisons work, so it kinda fits right in).

I don't suppose we _need_ to fix it as part of this (after all, it's already better than `master`), but we should at least add the test and know that the fix is incomplete, so we can either revisit it or keep its behavior stable.